File: tests/archive/ad_hoc/test_qt_arc.py
import sys
from PyQt5.QtWidgets import QApplication, QWidget
from PyQt5.QtGui import QPainter, QPen, QColor
from PyQt5.QtCore import Qt, QRectF
import math

class W(QWidget):
    def paintEvent(self, e):
        p = QPainter(self)
        p.translate(100, 100)
        
        KICAD_SCALE = 10.0
        x1 = 0 * KICAD_SCALE
        y1 = -(-1.524) * KICAD_SCALE
        x2 = -0.508 * KICAD_SCALE
        y2 = -(-1.902) * KICAD_SCALE
        x3 = -1.016 * KICAD_SCALE
        y3 = -(-1.524) * KICAD_SCALE

        D = 2 * (x1*(y2 - y3) + x2*(y3 - y1) + x3*(y1 - y2))
        cx = ((x1*x1 + y1*y1)*(y2 - y3) + (x2*x2 + y2*y2)*(y3 - y1) + (x3*x3 + y3*y3)*(y1 - y2)) / D
        cy = ((x1*x1 + y1*y1)*(x3 - x2) + (x2*x2 + y2*y2)*(x1 - x3) + (x3*x3 + y3*y3)*(x2 - x1)) / D
        r = math.hypot(x1 - cx, y1 - cy)

        ang1 = math.degrees(math.atan2(cy - y1, x1 - cx))
        ang2 = math.degrees(math.atan2(cy - y2, x2 - cx))
        ang3 = math.degrees(math.atan2(cy - y3, x3 - cx))
        
        if ang1 < 0: ang1 += 360.0
        if ang2 < 0: ang2 += 360.0
        if ang3 < 0: ang3 += 360.0

        ccw12 = ang2 - ang1
        if ccw12 < 0: ccw12 += 360.0
        ccw23 = ang3 - ang2
        if ccw23 < 0: ccw23 += 360.0
        cw12  = ang1 - ang2
        if cw12 < 0: cw12 += 360.0
        cw23  = ang2 - ang3
        if cw23 < 0: cw23 += 360.0

        span = 0
        if (ccw12 + ccw23) < 360.0:
            span = ccw12 + ccw23
        else:
            span = -(cw12 + cw23)

        rect = QRectF(cx - r, cy - r, 2*r, 2*r)
        p.setPen(QPen(Qt.red, 2))
        p.drawArc(rect, int(ang1 * 16), int(span * 16))
        
app = QApplication(sys.argv)
w = W()
w.resize(200, 200)
pix = w.grab()
pix.save("test_qt_arc.png")

File: tests/archive/ad_hoc/test_qt_arc.cpp
#include <QApplication>
#include <QPainter>
#include <QPixmap>
#include <iostream>
#include <cmath>

int main(int argc, char *argv[]) {
    QApplication app(argc, argv);
    QPixmap pix(200, 200);
    pix.fill(Qt::white);
    QPainter p(&pix);
    p.translate(100, 100);
    p.setPen(QPen(Qt::red, 2));

    double cx = -5.08, cy = 13.71;
    double r = 5.303;

    double ang1 = -16.8;    // Qt mathematical angle mapping
    double ang2 = -90.0;
    double ang3 = -163.2;

    if (ang1 < 0) ang1 += 360.0;
    if (ang2 < 0) ang2 += 360.0;
    if (ang3 < 0) ang3 += 360.0;

    double span = -146.6;

    QRectF rect(cx - r, cy - r, 2*r, 2*r);
    p.drawArc(rect, int(ang1 * 16.0), int(span * 16.0));
    
    // Test the second arc
    double cx2 = 5.08, cy2 = 16.51;
    double r2 = 5.236;
    double ang1_2 = -(-16.0); // 16.0
    double ang2_2 = -(-90.0); // 90.0
    double ang3_2 = -( -164.0); // 164.0

    if (ang1_2 < 0) ang1_2 += 360.0;
    if (ang2_2 < 0) ang2_2 += 360.0;
    if (ang3_2 < 0) ang3_2 += 360.0;
    
    double span2 = 148.0;
    
    QRectF rect2(cx2 - r2, cy2 - r2, 2*r2, 2*r2);
    p.setPen(QPen(Qt::blue, 2));
    p.drawArc(rect2, int(ang1_2 * 16.0), int(span2 * 16.0));

    p.end();
    pix.save("qt_arc_test.png");
    std::cout << "Done! Bounding Boxes:\n";
    std::cout << "1: " << rect.x() << " " << rect.y() << " " << rect.width() << " " << rect.height() << "\n";
    std::cout << "2: " << rect2.x() << " " << rect2.y() << " " << rect2.width() << " " << rect2.height() << "\n";
    return 0;
}
